draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,281 @@ from "have the same registrable domain" (or "is an exact match for") to say
+   sets a cookie `foo=http`. Previously this would result in one of
+   the cookies being overwritten by the other. Now, they're considered
+   seperate and when visiting each site users could see their
+   respective cookie. The same holds true for `https://example.com` and
+   `https://example.com:444`, each is a different origin and thus have
+   their own cookies.
+
+It's possible that a site operator purposefully wants some cookies to
+be accessible across port boundaries. Because this would weaken the
+origin boundary, by increasing the cookie's scope, we'd want this to be
+an opt-in mechanism. Helpfully there is already an attribute that
+increases a cookie's scope: the `Domain` attribute. We can modify the
+`Domain` attribute slightly such that it not only allows cookies to be
+accessible by different hosts but also by different ports than the
+origin's that set it. For convenience we'll call any cookie with a
+`Domain` attribute a "domain cookie".
+
+Example:
+
+* A corporate network, `https://corp.example`, has various services
+  each running on their own port. Those services share use of a token
+  which allows users to use the services. To get a token a user logs
+  in on `https://corp.example/login/` which creates the cookie
+  `IDToken=a1b2bc3`. Next, to file an expense, the user needs to
+  visit `https://corp.example:8443`, but when they do the IDToken
+  isn't sent because the ports differ (443 vs 8443) and the user is
+  denied access. To remedy this, the IDToken can have the `Domain`
+  attribute added to it: `IDToken=a1b2d3; Domain=corp.example`.
+  Now when the user visits `https://corp.example:8443` the token
+  is sent and access is granted as expected.
+
+Because any domain cookie is now exposed to multiple origins it means
+a cookie created by one origin can be overwritten by another origin. 
+
+Example:
+
+* `https://example.com:123` sets a cookie: `Set-Cookie:
+  foo=domaincookie; Domain=example.com`. The user then visits
+  `https://example.com:456` which sets the same cookie. This second
+  cookie would match the first and then overwrite it.
+    
+Also, because domain cookies are less trusted due to their wider scope,
+we'll want to avoid them shadowing non-domain cookies by disallowing
+domain cookies from being sent if a matching non-domain cookie exists.
+This is a departure from the status quo in which this shadowing
+behavior is specifically allowed.
+
+Example:
+
+* `https://example.com:456` sets a non-domain cookie `Set-Cookie:
+  foo=origincookie`. The user then visits `https://example.com:123`
+  which sets a domain cookie `Set-Cookie: foo=domaincookie;
+  Domain=example.com`. When the user returns to
+  `https://example.com:456` only `foo=origincookie` is sent and the
+  domain cookie is blocked from being sent. If the user were to then
+  visit `https://sub.example.com:456` `foo=domaincookie` would be sent
+  since the domain would no longer be shadowing a non-domain cookie.
```

Comment:
I mentioned `path` in some other thread that I can't find at the moment. I still think it's probably worth mentioning in prose somewhere. That is, consider `https://example.com:456/` setting `foo=origincookie; path=/whatever`. Does `foo=domaincookie` get sent on requests to paths other than`https://example.com:456/whatever`? I think it doesn't, based on my skim of the algorithm below, but spelling that out would be helpful.

Reply:
You're correct. I've added another example.